### PR TITLE
Pass siblings

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
This PR addresses #46 by passing a `siblings` keyword argument to the `check` function of subgraders in a `ListGrader`. This parameter is only available when `ordered=true`.

Currently only FormulaGrader makes any use of `siblings` in its check function. Example:
```python
grader=ListGrader(
    answers=['sibling_3 + 1', 'sibling_1^2', 'x'],
    subgraders=FormulaGrader(variables=['x']),
    ordered=True
)
grader(None, ['x + 1', 'x^2 + 2*x + 1', 'x'])
```
Some notes:

- This is a refactoring of #65, see discussion there.
- No author-facing documentation or course examples, though I've added that to the list of required documentation in #78 
- I've changed the sibling variable names from `input_k` to `sibling_k` to stress that the relationship is relative; particularly relevant when used with grouping / nested ListGraders.

### CODE REVIEW

File: mitxgraders/baseclasses.py [hunk 1]
```diff
@@ -92,13 +95,15 @@ def schema_config(self):
...
     @abc.abstractmethod
-    def check(self, answers, student_input):
+    def check(self, answers, student_input, **kwargs):
         """
         Check student_input for correctness and provide feedback.
 
         Arguments:
             answers: The expected result(s) and grading information
             student_input: The student's input passed by edX
+            **kwargs: Anything else that has been passed in For example, sibling
```
Comment (jolyonb) [inline mitxgraders/baseclasses.py]: Missing period before For.

File: mitxgraders/baseclasses.py [hunk 2]
```diff
@@ -92,13 +95,15 @@ def schema_config(self):
...
+                graders when a grader is used as a subgrader ing a ListGrader.
```
Comment (jolyonb) [inline mitxgraders/baseclasses.py]: ing?

File: mitxgraders/baseclasses.py [hunk 3]
```diff
@@ -298,14 +303,16 @@ def validate_expect(expect):
...
-    def check(self, answers, student_input):
+    def check(self, answers, student_input, **kwargs):
         """
         Compares student input to each answer in answers, using check_response.
         Computes the best outcome for the student.
 
         Arguments:
             answer: A tuple of answers to compare to, or None to use internal config
             student_input (str): The student's input passed by edX
+            **kwargs: Anything else that has been passed in For example, sibling
```
Comment (jolyonb) [inline mitxgraders/baseclasses.py]: Copy paste errors from above.

File: mitxgraders/baseclasses.py [hunk 4]
```diff
@@ -347,4 +354,6 @@ def check_response(self, answer, student_input):
         Arguments:
             answer (schema_answer): The answer to compare to
             student_input (str): The student's input passed by edX
+            **kwargs: Anything else that has been passed in For example, sibling
```
Comment (jolyonb) [inline mitxgraders/baseclasses.py]: More copy paste errors

File: mitxgraders/formulagrader.py [hunk 5]
```diff
@@ -613,39 +639,76 @@ def raw_check(self, answer, student_input):
...
+                key: scoped_eval(sibling_formulas[key])[0]
+                for key in sibling_formulas
+            }
+            varlist.update(siblings_eval)
+
+            # Compute expressions
+            comparer_params_eval = self.eval_and_validate_comparer_params(
+                scoped_eval, answer['expect']['comparer_params'], siblings_eval)
+
+            # Before performing student evaluation, scrub the dependencies
```
Comment (jolyonb) [inline mitxgraders/formulagrader.py]: dependencies -> siblings

File: mitxgraders/formulagrader.py [hunk 6]
```diff
@@ -613,39 +639,76 @@ def raw_check(self, answer, student_input):
...
+            # so that students can't use them
+            for key in siblings_eval:
+                del varlist[key]
+
+            student_eval, used = scoped_eval(student_input)
+
             # Check if expressions agree
             comparer_result = comparer(comparer_params_eval, student_eval, self.comparer_utils)
             if self.config['debug']:
+                # Put the dependencies back in for the debug output
```
Comment (jolyonb) [inline mitxgraders/formulagrader.py]: dependencies -> siblings

File: mitxgraders/formulagrader.py [hunk 7]
```diff
@@ -613,39 +639,76 @@ def raw_check(self, answer, student_input):
...
+
+    @staticmethod
+    def eval_and_validate_comparer_params(scoped_eval, comparer_params, siblings_eval):
+        """
+        Evaluate the comparer_params, and make sure they contain no references
+        to empty siblings.
+
+        Arguments
+        =========
+        - scoped_eval (func): a unuary function to evaluate math expressions.
```
Comment (jolyonb) [inline mitxgraders/formulagrader.py]: unary

File: mitxgraders/helpers/calc.py [hunk 8]
```diff
@@ -170,6 +171,8 @@ def get_parser(self, formula, suffixes):
 # The global parser cache
 parsercache = ParserCache()
 
+ScopeUsage = namedtuple('ScopeUSage', ['variables', 'functions', 'suffixes'])
```
Comment (jolyonb) [inline mitxgraders/helpers/calc.py]: I like this!

File: mitxgraders/listgrader.py [hunk 9]
```diff
@@ -446,6 +447,27 @@ def validate_submission(self, answers, student_list):
                 msg = "The number of answers ({}) and the number of inputs ({}) are different"
                 raise ConfigError(msg.format(len(answers), len(student_list)))
 
+    def get_ordered_input_list(self, answers, grouped_inputs):
+        """
+        Pass answers and inputs to the appropriate grader, along with sibling
+        information.
+        """
+        # If 'subgradres' is a single grader, create a list of references to it.
```
Comment (jolyonb) [inline mitxgraders/listgrader.py]: subgraders

**Comment by ChristopherChudzicki** [comment]
@jolyonb Cool, I'll fix those typos!

One of the things on your checklist for #65 was validation. Two comments about this:

1. I suspect the most common configuration error would be something like:
```python
grader = listGrader(
  answers=['1', 'sibling_1 + 1'],
  subgraders=FormulaGrader()
) # instantiated OK
grader(None, ['1', '2']) # throws InvalidInput: sibling_1 undefined
```
The issue is that the author forgot to declare the ListGrader as being ordered. FormulaGrader validates itself without any knowledge of sibling variables or the ListGrader's answers, so the error doesn't arise until check time.
  - This is sort of present in your original implementation (same error if author tries to use `input_k` variables without declaring `dependent_inputs`). You avoided number 2 by checking for cyclic dependencies in `dependent_inputs`.

2. Authors can set up ridiculous things like
```python
grader = ListGrader(
    answers=['sibling_1 + 1'],
    subgraders=FormulaGrader(),
    ordered=True
)
```
Nothing would ever be marked correct, since the answer cannot be 1 bigger than itself.
  - I am not very worried about this issue.

Mostly: **I think avoiding the extra key (`dependent_inputs`) and allowing reuse of the subgrader is worth the trouble of these less-than-helpful error messages.**

I'm also not super worried about it for what I suspect will be an uncommonly used feature.

A hacky but fast resolution to the first issue would be to raise a special error if the undefined variable contains `'siblings'`. Students *might* see this, but why would they be entering 'sibling_k' as a variable name....

Anyway, I think the PR is fine, but I wanted you to be aware of those issues.

**Comment by ChristopherChudzicki** [comment]
@jolyonb Typos fixed. I need to figure out a good way to spellcheck my code in Atom... sorry about that :)

**Comment by jolyonb** [comment]
Ah, I responded to your comment in #65...

**Pull request merged**